<a href="https://colab.research.google.com/github/JaneOrih/data-science-journey/blob/main/Jane_Feature_Engineering_and_Cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
df= pd.read_csv('C:/Users/Admin/Downloads/preprocessed-data.csv')

In [ ]:
df['date']=pd.to_datetime(df['date'])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64221 entries, 0 to 64220
Data columns (total 18 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   date              64221 non-null  datetime64[ns]
 1   State             64221 non-null  object        
 2   Local_Government  64221 non-null  object        
 3   Market_Name       64221 non-null  object        
 4   latitude          64221 non-null  float64       
 5   longitude         64221 non-null  float64       
 6   category          64221 non-null  object        
 7   commodity         64221 non-null  object        
 8   unit              64221 non-null  object        
 9   priceflag         64221 non-null  object        
 10  pricetype         64221 non-null  object        
 11  currency          64221 non-null  object        
 12  price             64221 non-null  float64       
 13  usdprice          64221 non-null  float64       
 14  year              6422

# Treating Outlier by price of each Commodity

In [ ]:
# Average price based on commodity and unit of food items
df.groupby(['commodity', 'unit'])['price'].mean()

commodity                    unit      
Bananas                      1KG              16.263438
Beans (niebe)                1KG             252.294500
Beans (red)                  1KG              38.065116
Beans (white)                1KG              34.919735
Bread                        Unit            264.564173
Cassava meal (gari, yellow)  1KG             229.160238
Cowpeas                      1KG              27.399075
Cowpeas (brown)              1KG             313.359220
Cowpeas (white)              1KG             284.857645
Eggs                         30 pcs         1217.207886
Fish                         1KG              30.154625
Fuel (diesel)                L               281.256098
Fuel (petrol-gasoline)       L               153.291524
Gari (white)                 1KG             195.490945
Groundnuts                   1KG              41.877470
Groundnuts (shelled)         1KG             368.037265
Maize                        1KG             172.553387
Maize (w

In [ ]:
# Grouping the dataframe into different dataframes based on commodity and unit

lst_var_df=[]

for variables in df['commodity'].unique():
    if len(df[df['commodity']== variables]['unit'].unique()) ==1:
        var_df= df[df['commodity']== variables]
        lst_var_df.append(var_df)
    else:
        var_df= df[df['commodity']== variables]
        for units in var_df['unit'].unique():
            var_dfs=var_df[var_df['unit']==units]
            lst_var_df.append(var_dfs)


In [ ]:
# outlier removal from each dataframe

no_outlier_df= []

for data in lst_var_df:
    df_sub= pd.DataFrame(data['price'])
    Q1= df_sub.quantile(0.25)
    Q3= df_sub.quantile(0.75)
    IQR = Q3 - Q1
    cleaned_df = df_sub[~((df_sub < (Q1 - 1.5 * IQR)) |(df_sub > (Q3 + 1.5 * IQR))).any(axis=1)]
    no_outlier_df.append(cleaned_df)

In [ ]:
#collecting all the index of leftover rows after outliers are removed

no_outlier_indices=[]
for x in range(len(no_outlier_df)):
    index_list=no_outlier_df[x].index.values.tolist()
    no_outlier_indices.append(index_list)


In [ ]:
# proper list of all index needed

indices=[]
for i in range(len(no_outlier_indices)):
    indices += no_outlier_indices[i]

len(indices)


61869

In [ ]:
len(df)

64221

After removing all ouliers, 61869 rows were remaining, out of 64221 rows

In [ ]:
# new dataframe
no_outlier_df= df.iloc[indices]